From dae9afd6dca4cf6f78088222b2ced3f0e4f17996 Mon Sep 17 00:00:00 2001
From: Zheng Zengkai <zhengzengkai@huawei.com>
Date: Sat, 29 Oct 2022 15:12:43 +0800
Subject: [PATCH 0642/1646] [Backport] x86: Avoid kabi change caused by adding
 pkru element in thread_struct

Intel inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/intel-kernel/issues/I5RQLJ

Reference: https://gitee.com/openeuler/kernel/commit/ad307c6f2109b6e11415665edb3700c58c63a136

--------------------------------

Assuming that struct thread_struct won't be accessed directly by drivers,
Using KABI_BROKEN_INSERT macro to avoid kabi change caused by commit 9782a712eb97
("x86/fpu: Add PKRU storage outside of task XSAVE buffer").

Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Lin Wang <lin.x.wang@intel.com>
Signed-off-by: Aichun Shi <aichun.shi@intel.com>
Signed-off-by: Jialin Zhang <zhangjialin11@huawei.com>
---
 arch/x86/include/asm/processor.h | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/x86/include/asm/processor.h b/arch/x86/include/asm/processor.h
index 134ca2536455..083888208933 100644
--- a/arch/x86/include/asm/processor.h
+++ b/arch/x86/include/asm/processor.h
@@ -35,6 +35,7 @@ struct vm86;
 #include <linux/err.h>
 #include <linux/irqflags.h>
 #include <linux/mem_encrypt.h>
+#include <linux/kabi.h>
 
 /*
  * We handle most unaligned accesses in hardware.  On the other hand
@@ -550,6 +551,7 @@ struct thread_struct {
 	unsigned int		iopl_warn:1;
 	unsigned int		sig_on_uaccess_err:1;
 
+#ifdef CONFIG_X86_64
 	/*
 	 * Protection Keys Register for Userspace.  Loaded immediately on
 	 * context switch. Store it in thread_struct to avoid a lookup in
@@ -557,7 +559,8 @@ struct thread_struct {
 	 * task is scheduled out. For 'current' the authoritative source of
 	 * PKRU is the hardware itself.
 	 */
-	u32                     pkru;
+	KABI_BROKEN_INSERT(u32 pkru)
+#endif
 
 	/* Floating point and extended processor state */
 	struct fpu		fpu;
-- 
2.17.1

